src/Libraries/Nop.Services/Helpers/BrowscapXmlParser.cs

```diff
@@ -101,7 +104,11 @@ private static string ToRegexp(string str)
         /// <returns>True if user agent is a crawler, otherwise - false</returns>
         public bool IsCrawler(string userAgent)
         {
-            return _crawlerUserAgentsRegexp.Any(p => Regex.IsMatch(userAgent, p));
+            // check against cache object
+            bool result;
+            if (_userAgentsCache.TryGetValue(userAgent, out result))
```

Comment:
New logic updated in UserAgentHelper, uses a found cache with fallback to a cookie and then a final fallback to BrowscapXmlParser.